From b476b1287bf560bb568ad24e86c92f9d644f16e0 Mon Sep 17 00:00:00 2001
From: Leo Chang <schang@qca.qualcomm.com>
Date: Tue, 14 Jan 2014 21:47:24 -0800
Subject: [PATCH 1/2] wlan: Invalid reload driver fix

If firmware could not power collapse whatever reason,
DXE interrupt could not be cleared if ISR is scheduled when IMPS mode.
Then next firmware power restore,
Empty interrupt will be issued when host power state change.
Incase, IMPS mode received empty interrupt,
and firmware could not power collapse,
host will detect this situation as successive empty interrupt.
Then issue SSR.
To prevent this invalid SSR, if interrupt happen with IMPS mode,
set dummy frame count to DXE channel control context.

Change-Id: I033d8a0cc4101aaa2d2954217dfa34cfd15e6562
CRs-fixed: 599833

Conflicts:
	prima/CORE/DXE/src/wlan_qct_dxe.c
---
 CORE/DXE/src/wlan_qct_dxe.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/CORE/DXE/src/wlan_qct_dxe.c b/CORE/DXE/src/wlan_qct_dxe.c
index 0febc3b..0fa5712 100644
--- a/CORE/DXE/src/wlan_qct_dxe.c
+++ b/CORE/DXE/src/wlan_qct_dxe.c
@@ -2686,6 +2686,13 @@ void dxeRXEventHandler
          HDXE_MSG(eWLAN_MODULE_DAL_DATA, eWLAN_PAL_TRACE_LEVEL_ERROR,
                   "dxeRXEventHandler Pull from RX high channel fail");        
       }
+      /* In case FW could not power collapse in IMPS mode
+       * Next power restore might have empty interrupt
+       * If IMPS mode has empty interrupt since RX thread race,
+       * Invalid re-load driver might happen
+       * To prevent invalid re-load driver,
+       * IMPS event handler set dummpy frame count */
+      channelCb->numFragmentCurrentChain = 1;
 
        /* Second low priority */
       channelCb = &dxeCtxt->dxeChannel[WDTS_CHANNEL_RX_LOW_PRI];
@@ -2697,6 +2704,8 @@ void dxeRXEventHandler
          HDXE_MSG(eWLAN_MODULE_DAL_DATA, eWLAN_PAL_TRACE_LEVEL_ERROR,
                   "dxeRXEventHandler Pull from RX low channel fail");        
       }
+      /* LOW Priority CH same above */
+      channelCb->numFragmentCurrentChain = 1;
 
       /* Interrupt will not enabled at here, it will be enabled at PS mode change */
       tempDxeCtrlBlk->rxIntDisabledByIMPS = eWLAN_PAL_TRUE;
-- 
1.8.2.1

